// @utils/src/MainConfig.ts
import path from "path";
import { PathManager } from "./PathManager";

/**
 * 全局配置
 */
export class MainConfig {
    /** 名字，最好只包含字母和数字 */
    static Name = 'linkage';
    /** 中文名字 */
    static ZHName = '联合';
    /** 版本 */
    static Version = '1.0.0';
    /** 服务端配置 */
    static server = {
        /** 端口 */
        port: 6102,
        /** 暗号相关 */
        secretCode: {
            /** 是否验证 */
            v: false,
            /** 过期时间 */
            overrunTime: 60 * 1000,
        },
        /** 
         * 阿里云相关配置
         * TODO 测试用的临时配置
         */
        aliOss: {
            region: "oss-cn-hangzhou",
            bucket: "yayaluoya-test",
            access: {
                accessKeyId: "LTAI5tJmDzkUiD911YUVDjMa",
                accessKeySecret: "MtdSYoxiHEaioWff2M9v4KafU5XBj4",
            },
            roleArn: 'acs:ram::1330807560356268:role/yayaluoya-test',
        },
        /** 公共文件缓存时间 */
        publicFileMaxAge: 1000 * 60 * 60 * 24 * 360,
        /** 是否运行测试模块 */
        ifTest: true,
    };
    /** 关于orm的配置 */
    static orm = {
        type: "mysql",
        host: "localhost",
        port: 3306,
        username: "root",
        password: "123456",
        database: "test",
        entities: [
            path.join(PathManager.server, "./dist/db/entities/**/*{.ts,.js}")
        ],
        synchronize: true
    };
}
